ctlra/devices/midi_generic.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2017, OpenAV Productions,
+ * Harry van Haaren <harryhaaren@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+ * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "impl.h"
+#include "midi.h"
+
+#define CONTROLS_SIZE 512
+#define LIGHTS_SIZE 512
+
+#define AKAI    0x09e8
+#define APC40   0x0073
+#define MPKMINI 0x007c
+
+#define CTLRA_DRIVER_VENDOR AKAI
```

Comment:
@harryhaaren Do these make sense here? (since it's called `midi_generic`)

Reply:
Good point. No they don't make sense. One thing that this PR doesn't address yet is how to "discover" or "map" generic MIDI controller devices. Currently the code uses this as a hacky method to register the driver.. but this should be refactored to allow more `midi_generic` type uses.

Right now, I don't have a good handle on what it should be refactored _to_, and hence no progress has been made in that direction yet.. but it is intended to be generic. Currently its not. Thanks for pointing that out clearly - pin-pointing the issue is the first step to solving! :)  -H

Reply:
Removed device specific defines in commit below. Thanks for review!